File: ts/src/algorithms/linear-regression.ts
/**
 * Linear Regression for time series forecasting
 *
 * @module algorithms/linear-regression
 */

import { getWasmModule, ensureWasm } from '../wasm-loader';
import type { Predictor, TimeSeriesData } from '../types';

/**
 * Linear Regression model for time series
 *
 * Fits y = intercept + slope * t where t is the time index.
 * Good for data with a clear linear trend.
 *
 * @example
 * ```typescript
 * import { initWasm, LinearRegression } from 'rustful-ts';
 *
 * await initWasm();
 *
 * const model = new LinearRegression();
 * await model.fit([10, 12, 14, 16, 18, 20]);
 *
 * console.log(model.getSlope());     // ~2.0
 * console.log(model.getIntercept()); // ~10.0
 * console.log(model.getRSquared());  // ~1.0
 *
 * const forecast = await model.predict(3);
 * // Returns [22, 24, 26]
 * ```
 */
export class LinearRegression implements Predictor {
  private inner: unknown = null;
  private fitted = false;

  constructor() {}

  async fit(data: TimeSeriesData): Promise<void> {
    await ensureWasm();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const wasm = getWasmModule() as any;

    this.inner = new wasm.WasmLinearRegression();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (this.inner as any).fit(new Float64Array(data));
    this.fitted = true;
  }

  async predict(steps: number): Promise<number[]> {
    if (!this.fitted || !this.inner) {
      throw new Error('Model must be fitted before prediction');
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result = (this.inner as any).predict(steps);
    return Array.from(result as Float64Array);
  }

  isFitted(): boolean {
    return this.fitted;
  }

  /**
   * Get the slope (trend per time unit)
   */
  getSlope(): number {
    if (!this.fitted || !this.inner) {
      throw new Error('Model must be fitted first');
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (this.inner as any).slope();
  }

  /**
   * Get the y-intercept
   */
  getIntercept(): number {
    if (!this.fitted || !this.inner) {
      throw new Error('Model must be fitted first');
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (this.inner as any).intercept();
  }

  /**
   * Get R-squared (coefficient of determination)
   *
   * Values close to 1.0 indicate a good fit.
   */
  getRSquared(): number {
    if (!this.fitted || !this.inner) {
      throw new Error('Model must be fitted first');
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (this.inner as any).r_squared();
  }
}

/**
 * Linear Regression with seasonal dummy variables
 *
 * Extends basic linear regression to capture seasonal patterns.
 *
 * @example
 * ```typescript
 * import { initWasm, SeasonalLinearRegression } from 'rustful-ts';
 *
 * await initWasm();
 *
 * // Monthly data with yearly seasonality
 * const model = new SeasonalLinearRegression(12);
 * await model.fit(monthlyData); // At least 24 observations
 *
 * const forecast = await model.predict(12);
 * ```
 */
export class SeasonalLinearRegression implements Predictor {
  private fitted = false;
  private readonly period: number;
  private slope = 0;
  private intercept = 0;
  private seasonalFactors: number[] = [];
  private nObservations = 0;

  /**
   * Create a new seasonal linear regression model
   *
   * @param period - Number of observations per seasonal cycle
   */
  constructor(period: number) {
    if (period < 2) {
      throw new Error('Period must be at least 2');
    }
    this.period = period;
  }

  async fit(data: TimeSeriesData): Promise<void> {
    const minRequired = this.period * 2;
    if (data.length < minRequired) {
      throw new Error(`Need at least ${minRequired} observations`);
    }

    this.nObservations = data.length;

    // Step 1: Fit linear trend
    const n = data.length;
    const sumT = (n * (n - 1)) / 2;
    const sumY = data.reduce((a, b) => a + b, 0);
    const sumT2 = (n * (n - 1) * (2 * n - 1)) / 6;
    const sumTY = data.reduce((acc, y, t) => acc + t * y, 0);

    const denominator = n * sumT2 - sumT * sumT;
    this.slope = (n * sumTY - sumT * sumY) / denominator;
    this.intercept = (sumY - this.slope * sumT) / n;

    // Step 2: Detrend
    const detrended = data.map((y, t) => y - (this.intercept + this.slope * t));

    // Step 3: Compute seasonal factors
    this.seasonalFactors = new Array(this.period).fill(0);
    const counts = new Array(this.period).fill(0);

    detrended.forEach((val, i) => {
      const seasonIdx = i % this.period;
      this.seasonalFactors[seasonIdx] += val;
      counts[seasonIdx]++;
    });

    for (let i = 0; i < this.period; i++) {
      if (counts[i] > 0) {
        this.seasonalFactors[i] /= counts[i];
      }
    }

    // Normalize to sum to 0
    const meanFactor =
      this.seasonalFactors.reduce((a, b) => a + b, 0) / this.period;
    this.seasonalFactors = this.seasonalFactors.map((f) => f - meanFactor);

    this.fitted = true;
  }

  async predict(steps: number): Promise<number[]> {
    if (!this.fitted) {
      throw new Error('Model must be fitted before prediction');
    }

    const forecasts: number[] = [];
    for (let i = 0; i < steps; i++) {
      const t = this.nObservations + i;
      const seasonIdx = t % this.period;
      const trend = this.intercept + this.slope * t;
      forecasts.push(trend + this.seasonalFactors[seasonIdx]);
    }

    return forecasts;
  }

  isFitted(): boolean {
    return this.fitted;
  }

  /**
   * Get seasonal factors
   */
  getSeasonalFactors(): number[] {
    if (!this.fitted) {
      throw new Error('Model must be fitted first');
    }
    return [...this.seasonalFactors];
  }

  /**
   * Get the slope
   */
  getSlope(): number {
    return this.slope;
  }

  /**
   * Get the intercept
   */
  getIntercept(): number {
    return this.intercept;
  }
}

File: crates/rustful-core/src/algorithms/linear_regression.rs
//! Linear Regression for time series forecasting
//!
//! Uses ordinary least squares (OLS) to fit a linear trend to time series data.
//! Can be extended with seasonal dummy variables for seasonal decomposition.
//!
//! ## When to Use
//!
//! - Data shows a clear linear trend
//! - Quick baseline model
//! - When interpretability is important

use crate::algorithms::Predictor;
use crate::error::{Result, TsError};
use serde::{Deserialize, Serialize};

/// Linear Regression model for time series
///
/// Fits y = intercept + slope * t where t is the time index.
///
/// # Example
///
/// ```rust
/// use rustful_ts::algorithms::linear_regression::LinearRegression;
/// use rustful_ts::algorithms::Predictor;
///
/// let data = vec![10.0, 12.0, 14.0, 16.0, 18.0, 20.0];
/// let mut model = LinearRegression::new();
/// model.fit(&data).unwrap();
///
/// let forecast = model.predict(3).unwrap();
/// // Should predict approximately [22, 24, 26]
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearRegression {
    /// Y-intercept
    intercept: f64,
    /// Slope (trend per time unit)
    slope: f64,
    /// Number of observations used in fitting
    n_observations: usize,
    /// R-squared value
    r_squared: f64,
    /// Whether model has been fitted
    fitted: bool,
}

impl Default for LinearRegression {
    fn default() -> Self {
        Self::new()
    }
}

impl LinearRegression {
    /// Create a new linear regression model
    pub fn new() -> Self {
        Self {
            intercept: 0.0,
            slope: 0.0,
            n_observations: 0,
            r_squared: 0.0,
            fitted: false,
        }
    }

    /// Get the slope (trend per time unit)
    pub fn slope(&self) -> f64 {
        self.slope
    }

    /// Get the intercept
    pub fn intercept(&self) -> f64 {
        self.intercept
    }

    /// Get R-squared (coefficient of determination)
    pub fn r_squared(&self) -> f64 {
        self.r_squared
    }

    /// Predict value at a specific time index
    pub fn predict_at(&self, t: f64) -> Result<f64> {
        if !self.fitted {
            return Err(TsError::NotFitted);
        }
        Ok(self.intercept + self.slope * t)
    }

    /// Get residuals from the fit
    pub fn residuals(&self, data: &[f64]) -> Vec<f64> {
        if !self.fitted {
            return Vec::new();
        }

        data.iter()
            .enumerate()
            .map(|(i, &y)| y - (self.intercept + self.slope * i as f64))
            .collect()
    }
}

impl Predictor for LinearRegression {
    fn fit(&mut self, data: &[f64]) -> Result<()> {
        if data.len() < 2 {
            return Err(TsError::InsufficientData {
                required: 2,
                actual: data.len(),
            });
        }

        let n = data.len() as f64;
        self.n_observations = data.len();

        // Time indices: 0, 1, 2, ...
        let sum_t: f64 = (0..data.len()).map(|i| i as f64).sum();
        let sum_y: f64 = data.iter().sum();
        let sum_t2: f64 = (0..data.len()).map(|i| (i * i) as f64).sum();
        let sum_ty: f64 = data.iter().enumerate().map(|(i, &y)| i as f64 * y).sum();

        // OLS formulas
        let denominator = n * sum_t2 - sum_t * sum_t;
        if denominator.abs() < 1e-10 {
            return Err(TsError::NumericalError(
                "Singular matrix in regression".to_string(),
            ));
        }

        self.slope = (n * sum_ty - sum_t * sum_y) / denominator;
        self.intercept = (sum_y - self.slope * sum_t) / n;

        // Calculate R-squared
        let mean_y = sum_y / n;
        let ss_tot: f64 = data.iter().map(|&y| (y - mean_y).powi(2)).sum();
        let ss_res: f64 = data
            .iter()
            .enumerate()
            .map(|(i, &y)| {
                let predicted = self.intercept + self.slope * i as f64;
                (y - predicted).powi(2)
            })
            .sum();

        self.r_squared = if ss_tot > 1e-10 {
            1.0 - ss_res / ss_tot
        } else {
            1.0
        };

        self.fitted = true;
        Ok(())
    }

    fn predict(&self, steps: usize) -> Result<Vec<f64>> {
        if !self.fitted {
            return Err(TsError::NotFitted);
        }

        let mut forecasts = Vec::with_capacity(steps);
        for i in 0..steps {
            let t = (self.n_observations + i) as f64;
            forecasts.push(self.intercept + self.slope * t);
        }

        Ok(forecasts)
    }

    fn is_fitted(&self) -> bool {
        self.fitted
    }
}

/// Linear Regression with seasonal dummy variables
///
/// Extends basic linear regression to capture seasonal patterns.
///
/// # Example
///
/// ```rust
/// use rustful_ts::algorithms::linear_regression::SeasonalLinearRegression;
/// use rustful_ts::algorithms::Predictor;
///
/// // Monthly data with yearly seasonality
/// let data: Vec<f64> = (0..36).map(|i| {
///     10.0 + (i as f64 * 0.5) + 5.0 * ((i % 12) as f64 / 12.0 * std::f64::consts::PI * 2.0).sin()
/// }).collect();
///
/// let mut model = SeasonalLinearRegression::new(12).unwrap();
/// model.fit(&data).unwrap();
/// let forecast = model.predict(12).unwrap();
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SeasonalLinearRegression {
    /// Seasonal period
    period: usize,
    /// Base trend model
    trend: LinearRegression,
    /// Seasonal factors (additive)
    seasonal_factors: Vec<f64>,
    /// Whether model has been fitted
    fitted: bool,
}

impl SeasonalLinearRegression {
    /// Create a new seasonal linear regression model
    ///
    /// # Arguments
    ///
    /// * `period` - Number of observations per seasonal cycle
    pub fn new(period: usize) -> Result<Self> {
        if period < 2 {
            return Err(TsError::InvalidParameter {
                name: "period".to_string(),
                reason: "must be at least 2".to_string(),
            });
        }

        Ok(Self {
            period,
            trend: LinearRegression::new(),
            seasonal_factors: vec![0.0; period],
            fitted: false,
        })
    }

    /// Get seasonal factors
    pub fn seasonal_factors(&self) -> &[f64] {
        &self.seasonal_factors
    }

    /// Get the underlying trend model
    pub fn trend(&self) -> &LinearRegression {
        &self.trend
    }
}

impl Predictor for SeasonalLinearRegression {
    fn fit(&mut self, data: &[f64]) -> Result<()> {
        let min_required = self.period * 2;
        if data.len() < min_required {
            return Err(TsError::InsufficientData {
                required: min_required,
                actual: data.len(),
            });
        }

        // Step 1: Fit trend
        self.trend.fit(data)?;

        // Step 2: Detrend
        let detrended: Vec<f64> = data
            .iter()
            .enumerate()
            .map(|(i, &y)| y - (self.trend.intercept + self.trend.slope * i as f64))
            .collect();

        // Step 3: Compute seasonal factors as mean of detrended values for each season
        for s in 0..self.period {
            let values: Vec<f64> = detrended
                .iter()
                .enumerate()
                .filter(|(i, _)| i % self.period == s)
                .map(|(_, &v)| v)
                .collect();

            self.seasonal_factors[s] = if !values.is_empty() {
                values.iter().sum::<f64>() / values.len() as f64
            } else {
                0.0
            };
        }

        // Normalize seasonal factors to sum to 0
        let mean_factor: f64 = self.seasonal_factors.iter().sum::<f64>() / self.period as f64;
        for factor in &mut self.seasonal_factors {
            *factor -= mean_factor;
        }

        self.fitted = true;
        Ok(())
    }

    fn predict(&self, steps: usize) -> Result<Vec<f64>> {
        if !self.fitted {
            return Err(TsError::NotFitted);
        }

        let trend_forecasts = self.trend.predict(steps)?;

        let forecasts: Vec<f64> = trend_forecasts
            .iter()
            .enumerate()
            .map(|(i, &trend_val)| {
                let season_idx = (self.trend.n_observations + i) % self.period;
                trend_val + self.seasonal_factors[season_idx]
            })
            .collect();

        Ok(forecasts)
    }

    fn is_fitted(&self) -> bool {
        self.fitted
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_linear_regression() {
        let data: Vec<f64> = (0..10).map(|i| 10.0 + 2.0 * i as f64).collect();
        let mut model = LinearRegression::new();
        model.fit(&data).unwrap();

        assert!((model.slope() - 2.0).abs() < 1e-10);
        assert!((model.intercept() - 10.0).abs() < 1e-10);
        assert!(model.r_squared() > 0.99);

        let forecast = model.predict(3).unwrap();
        assert!((forecast[0] - 30.0).abs() < 1e-10);
        assert!((forecast[1] - 32.0).abs() < 1e-10);
        assert!((forecast[2] - 34.0).abs() < 1e-10);
    }

    #[test]
    fn test_seasonal_linear_regression() {
        let data: Vec<f64> = (0..24)
            .map(|i| 10.0 + 0.5 * i as f64 + [2.0, -1.0, 0.0, -1.0][i % 4])
            .collect();

        let mut model = SeasonalLinearRegression::new(4).unwrap();
        model.fit(&data).unwrap();

        let forecast = model.predict(4).unwrap();
        assert_eq!(forecast.len(), 4);
    }
}
